source/frontend/editors/RarcEditor.hpp

```diff
@@ -0,0 +1,299 @@
+      auto node_it = std::find(m_rarc.nodes.begin(), m_rarc.nodes.end(), node);
+      return node.folder.sibling_next -
+             std::distance(m_rarc.nodes.begin(), node_it) - 1;
+    }
+    return node.data.size();
+  }
+
+  void RenameNode(const librii::RARC::ResourceArchive::Node& node,
+                  const std::string& new_name) {
+    m_node_to_rename = node;
+    m_node_new_name = new_name;
+  }
+
+  void InsertFiles(const std::vector<rsl::File>& files,
+                   std::optional<s32> parent = std::nullopt) {
+    if (parent)
+      SetParentNode(*parent);
+    m_files_to_insert = files;
+  }
+
+  void InsertFolder(const std::filesystem::path& folder,
+                    std::optional<s32> parent = std::nullopt) {
+    if (parent)
+      SetParentNode(*parent);
+    m_folder_to_insert = folder;
+  }
+
+  void CreateFolder(const std::string& folder,
+                    std::optional<s32> parent = std::nullopt) {
+    if (parent)
+      SetParentNode(*parent);
+    m_folder_to_create = folder;
+  }
+
+  void ExtractNodeTo(const librii::RARC::ResourceArchive::Node& node,
+                     const std::filesystem::path& dst) {
+    m_node_to_extract = node;
+    m_extract_path = dst;
+  }
+
+  void ReplaceNodeWith(const librii::RARC::ResourceArchive::Node& node,
+                       const std::filesystem::path& dst) {
+    m_node_to_replace = node;
+    m_replace_path = dst;
+  }
+
+  void SetParentNode(s32 parent) { m_insert_parent = parent; }
+
+  void
+  DeleteNodes(const std::vector<librii::RARC::ResourceArchive::Node>& nodes) {
+    m_nodes_to_delete = nodes;
+  }
+
+  void FlagErrorDialog(const std::string& message) {
+    m_grid.m_operation_error_msg = message;
+    m_grid.m_operation_state = TriState::ST_FALSE;
+    m_grid.m_operation_modal_state = ModalState::M_OPENING;
+  }
+
+  void FlagErrorDialog(const std::error_code& err) {
+    m_grid.m_operation_error_msg = err.message();
+    m_grid.m_operation_state = TriState::ST_FALSE;
+    m_grid.m_operation_modal_state = ModalState::M_OPENING;
+  }
+
+  void FlagSuccessDialog() {
+    m_grid.m_operation_state = TriState::ST_TRUE;
+    m_grid.m_operation_modal_state = ModalState::M_OPENING;
+  }
+
+  void ResetOperationContexts() {
+    // Rename
+    m_node_to_rename = std::nullopt;
+    m_node_new_name = "";
+
+    // Insert
+    m_files_to_insert.clear();
+    m_folder_to_insert = std::nullopt;
+
+    // Create
+    m_folder_to_create = std::nullopt;
+
+    // Delete
+    m_nodes_to_delete.clear();
+
+    // Extract
+    m_node_to_extract = std::nullopt;
+
+    // Replace
+    m_node_to_replace = std::nullopt;
+  }
+
+  void ApplyUpdates() {
+    auto result = reconstruct();
+
+    ResetOperationContexts();
+
+    if (!result) {
+      FlagErrorDialog(result.error());
+    }
+  }
+
+private:
+  RarcEditorPropertyGrid m_grid;
+  RarcEditorTabSheet m_sheet;
+  std::string m_path;
+  librii::RARC::ResourceArchive m_rarc;
+
+  //// Edit state
+  bool m_changes_made;
```

Comment:
This will be undefined/randomly set on object construction unless a default value is specified. Should this be changed?

Reply:
Yes it should be changed. Simple oversight on my part.